loki/transform/transform_lift_contained_subroutines.py

```diff
@@ -0,0 +1,212 @@
+from loki import (
+    Subroutine, FindNodes, FindVariables,
+    CallStatement, VariableDeclaration, Import,
+    FindInlineCalls, Transformer, CaseInsensitiveDict,
+    DerivedType
+)
+from functools import reduce
+from loki.transform import resolve_associates
+
+def lift_contained_subroutines(routine):
```

Comment:
I introduced some minor modifications to also make it work for functions. The only difference is that for functions, need to transform `InlineCall`s in the parent instead of `CallStatement`s. I renamed everything to `procedures` instead of routines as well.I also added a two tests with functions. The test coverage for functions is not as good. However, it seems that functions are implemented using `Subroutine` (i.e they are very similar in Loki) so maybe this is not a problem? I can improve the tests if need be, the extreme of course being that each test is replicated such that the inner procedure is a function.